Payroll: update readme to latest changes and add more spec notes (#889)
aragon/aragon-apps: future-apps/payroll/README.md

- `ADD_REIMBURSEMENT_ROLE`: allows invocation of `addReimbursement()`
- `ALLOWED_TOKENS_MANAGER_ROLE`: allows invocation of `addAllowedToken()`
- `CHANGE_PRICE_FEED_ROLE`: allows invocation of `setPriceFeed()`
- `MODIFY_RATE_EXPIRY_ROLE`: allows invocation of `setRateExpiryTime()`

These roles can all be considered "management" roles, in that they are sensitive and meant to be held by more privileged members or apps in the organization. In particular, the set of `ADD_EMPLOYEE_ROLE`, `TERMINATE_EMPLOYEE_ROLE`, `SET_EMPLOYEE_SALARY_ROLE`, `ADD_BONUS_ROLE`, and `ADD_REIMBURSEMENT_ROLE` roles control functionality related to managing employees and their salaries, and the set of `ALLOWED_TOKENS_MANAGER_ROLE`, `CHANGE_PRICE_FEED_ROLE`, `MODIFY_RATE_EXPIRY_ROLE` roles control specific parameters and configuration points in how the Payroll app operates as a whole.

Employees themselves are authenticated based on their "employee account" (an address), and have access to functionality related to their account that only they are allowed to access:

- `changeAddressByEmployee()`
- `determineAllocation()`
- `payday()`

### Payroll management

> The Payroll app is [initialized](#initialization) with four parameters:

Reviewer: I think the `#initialization` link is missing

- **Finance**: a reference to a [Finance](https://github.com/aragon/aragon-apps/tree/master/apps/finance) instance that will be used to make payments
  - Finance in turn will use a [Vault](https://github.com/aragon/aragon-apps/tree/master/apps/vault) to access funds, but having the Payroll app use a Finance app allows salary payments to record more obvious transactions in the organization's UI

> - **Denomination token**: ["base"](https://www.investopedia.com/terms/b/basecurrency.asp) token used to denominate all salaries in this Payroll instance

Reviewer: ```suggestion
- **Denomination token**: ["base"](https://www.investopedia.com/terms/b/basecurrency.asp) token used to denominate all salaries in a Payroll instance
```

> - **Denomination token**: ["base"](https://www.investopedia.com/terms/b/basecurrency.asp) token used to denominate all salaries in this Payroll instance

Reviewer: ```suggestion
- **Denomination token**: ["base"](https://www.investopedia.com/terms/b/basecurrency.asp) token used to denominate all salaries in a Payroll instance
```

  - This denomination token may not be a real on-chain currency (depends on how the linked price feed is implemented and if the denomination token is allowed to be used as a salary token)
- **Price feed**: [`IFeed`](https://github.com/aragon/ppf/blob/master/packages/ppf-contracts/) implementation that provides exchange rates between different tokens

>   - May include exchange rates to/from "phantom" tokens, to mimic off-chain currencies (e.g. USD)

Reviewer: - This price feed is only used to query the exchange rate of the allowed tokens based on the denomination token set.
```suggestion
  - May include exchange rates to/from "phantom" tokens, to mimic off-chain currencies (e.g. USD)
  - This price feed is only used to query the exchange rate of the allowed tokens based on the denomination token set.
```

> - **Price feed expiration time**: Maximum latency allowed between an exchange rate's last update in the Price Feed and its actual use

Reviewer: We could state here that if a rate is considered "old" or expired, the requested payout will revert.

> tokens. You cannot yet revoke any allowed tokens.

### Employee management

### Employee actions

## Example lifecycle

### Payroll manager

1. **Installation**

Install an instance of the Payroll app onto the organization, and set up permissions in accordance to the governance structure desired.

For simplicity, we assume that all of the functionality discussed below can be achieved in one way or another (either one person directly holds the permission, or can access the functionality behind a forwarder).

2. **Initialization**

Initializing a Payroll app requires the following parameters:

- Finance
- Denomination token
- Price feed
- Price feed rate expiry time

See [Payroll management](#payroll-management) for more information on these specific parameters.

We assume the Payroll app has been instantiated using a Finance app installed onto the same organization and an already deployed price feed, using DAI as the denomination token and 10 minutes as the price feed rate expiry limit:

```
payroll.initialize(0xabc...def, 0x89d24a6b4ccb1b6faa2625fe562bdd9a23260359, 0xabc...def, 600);
```

3. **Add allowed tokens**

> Add ETH and DAI as allowed tokens for salary withdrawl:

Reviewer: ```suggestion
Add ETH and DAI as allowed tokens for salary withdrawal:
```

```
payroll.addAllowedToken(0x0000000000000000000000000000000000000000); // ETH
payroll.addAllowedToken(0x89d24a6b4ccb1b6faa2625fe562bdd9a23260359); // DAI
```

4. **Add employee**

Add an employee starting on Jan. 1st, 2020, with 80,000 DAI salary (note the salary is in seconds, and already adjusted for 18 decimals):

```
payroll.addEmployee(0xabc...def, 2535047025122317, 1577836800, "employee");
```

Assume for future examples this employee has an employee ID of `1`.

5. **Modify employee salary**

After some time, give the employee a raise by changing their salary to 90,000 DAI:

```
payroll.setEmployeeSalary(1, 2851927903262606);
```

6. **Terminate employee**

Finally, remove the employee from payroll on Jan. 1, 2022:

```
payroll.terminateEmployee(1, 1640995200);
```

### Employee

1. **Determine allocation**

```
payroll.determineAllocation(address[] tokens, uint8[] distribution)
```

Employees can set the proportion of every allowed token that want to be used for their salary payment. Distribution values are expressed as a ratio to 100.

2. **Request payroll**

```
payroll.payday()
```

> Employees can request payroll whenever they want and the proportional amount of their anual salary since the last request (or since the start date if it's the first one) will be transferred.

Reviewer: Should be `annual` I think

3. **Change account address**

```
payroll.changeAddressByEmployee(address newAddress)
```

Employees can change their own address.

4. **Claim final paycheck after termination**

```

> payroll.payday()

Reviewer: Current signature is `payday(PaymentType _type, uint256 _requestedAmount)`

```

## :warning: Gotchas to be aware of / potentially dangerous limitations

### Requirements For allowed and denomination tokens

> Due to the limits of mathematical precision, any token used as either the denonination token or an allowed token must have a "reasonable" exchange rate against all other allowed or denomination tokens and this exchange rate must be fetchable from the price feed.

Reviewer: ```suggestion
Due to the limits of mathematical precision, any token used as either the denomination token or an allowed token must have a "reasonable" exchange rate against all other allowed or denomination tokens and this exchange rate must be fetchable from the price feed.
```

Reviewer: I think we only care there is an exchange rate agains the denomination token right? I mean, we don't do any calculations between allowed tokens.

> A "resonable" exchange rate is one that falls between 1e-18 and 1e18 against a base token.

Reviewer: ```suggestion
A "reasonable" exchange rate is one that falls between 1e-18 and 1e18 against a base token.
```

### Lack of budget or funds

`payday()` fails if one of the token payments exceeds the budget left in the Finance app or the Vault's balance for that token. Even if the Vault could technically pay out the salary by converting its assets appropriately, no automatic conversions are attempted.

**Denial-of-service**: as salaries are not held in escrow, an organization could strategically transfer funds or move them into different accounts to deny employees' salaries.

> **Frontrunning**: in the event where the Finance app is not able to pay everyone's salary, only the first ones able to execute `payday()` will be paid.

Reviewer: ```suggestion
**Front-running**: in the event where the Finance app is not able to pay everyone's salary, only the first ones able to execute `payday()` will be paid.
```

### Inaccurate maths

> Due to the nature of exchange rates and running imprecise arithmetic operations, a user's actual, paid salary will differ slightly from their "real" salary. In all cases though, these rounding errors should be so ridiculuosly small as to not present a problem in reality.

Reviewer: ```suggestion
Due to the nature of exchange rates and running imprecise arithmetic operations, a user's actual, paid salary will differ slightly from their "real" salary. In all cases though, these rounding errors should be so ridiculously small as to not present a problem in reality.
```
